### PR TITLE
Added support for Organization Policy and Access Policy from CAI to Forseti inventory

### PR DESCRIPTION
**Pull request by red2k18**
This PR adds support for following policies from CAI.

Organization Policy
- [x] cloudresourcemanager.googleapis.com/Organization
- [x] cloudresourcemanager.googleapis.com/Folder
- [x] cloudresourcemanager.googleapis.com/Project

Access Policy
- [x] cloudresourcemanager.googleapis.com/Organization

This PR also adds support for Access Level and Service Perimeter
 - [x] https://cloud.google.com/access-context-manager/docs/reference/rest/v1/accessPolicies.accessLevels
- [x] https://cloud.google.com/access-context-manager/docs/reference/rest/v1/accessPolicies

This PR also adds support for all the above resources in the Model.

### CODE REVIEW

**Comment by blueandgold** [comment]
@joecheuk will do the review here, in conjunction with the VPC-SC rego work.

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py [hunk 1]
```diff
@@ -1148,6 +1148,74 @@ def iter_crm_projects(self, parent_type, parent_id):
         for project in resources:
             yield project
 
+    # pylint: disable=inconsistent-return-statements, using-constant-test
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Can you not use pylint disable here? The other methods don't seem to use it.

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py [hunk 2]
```diff
@@ -1148,6 +1148,74 @@ def iter_crm_projects(self, parent_type, parent_id):
...
+
+        Returns:
+            dict: Folder organization policy.
+        """
+        resource = self.dao.iter_cai_assets(
+            ContentTypes.org_policy,
+            'cloudresourcemanager.googleapis.com/Folder',
+            '//cloudresourcemanager.googleapis.com/{}'.format(folder_id),
+            self.engine)
+        if resource:
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: the if / return seems to be used in the fetch_..() method instead of the iter_..() method. Is there any reason why this method is using this pattern? e.g. method at line 1132 and 1219.

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py [hunk 3]
```diff
@@ -1148,6 +1148,74 @@ def iter_crm_projects(self, parent_type, parent_id):
...
+            return resource
+
+    # pylint: disable=inconsistent-return-statements, using-constant-test
+    def iter_crm_organization_access_policies(self, org_id):
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Inside access policies, there are sub resources (e.g. service_perimeter and access_level) under access_policies, too. Just wanted to make sure those are also included.
Example access policy in the dump file:
```
{"name":"//cloudresourcemanager.googleapis.com/organizations/660570133860","asset_type":"cloudresourcemanager.googleapis.com/Organization","access_policy":{"name":"accessPolicies/1008882730433","parent":"organizations/660570133860","title":"default policy"},"ancestors":["organizations/660570133860"]}
{"name":"//cloudresourcemanager.googleapis.com/organizations/660570133860","asset_type":"cloudresourcemanager.googleapis.com/Organization","service_perimeter":{"name":"accessPolicies/1008882730433/servicePerimeters/Test_Service_Parameter","title":"Test Service Parameter","status":{"resources":["projects/179891054368"],"access_levels":["accessPolicies/1008882730433/accessLevels/abcb"],"restricted_services":["container.googleapis.com"]}},"ancestors":["organizations/660570133860"]}
{"name":"//cloudresourcemanager.googleapis.com/organizations/660570133860","asset_type":"cloudresourcemanager.googleapis.com/Organization","access_level":{"name":"accessPolicies/1008882730433/accessLevels/abcb","title":"abcb","basic":{"conditions":[{"ip_subnetworks":["0.0.0.0/0"],"regions":["US"]}]}},"ancestors":["organizations/660570133860"]}
{"name":"//cloudresourcemanager.googleapis.com/organizations/660570133860","asset_type":"cloudresourcemanager.googleapis.com/Organization","access_level":{"name":"accessPolicies/1008882730433/accessLevels/Ebil","title":"Ebil","basic":{"conditions":[{"device_policy":{"require_screenlock":true,"allowed_encryption_statuses":[3]}}]}},"ancestors":["organizations/660570133860"]}
```

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 1 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Changing the return type allowed to remove pylint disables. Done.

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 2 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Updated to be consistent with other iter methods.

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 3 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Yes, it is included.

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py [hunk 4]
```diff
@@ -1148,6 +1148,110 @@ def iter_crm_projects(self, parent_type, parent_id):
...
+
+        Yields:
+            dict: Generator of access policies for an organization.
+        """
+        resources = self.dao.iter_cai_assets(
+            ContentTypes.access_policy,
+            'cloudresourcemanager.googleapis.com/Organization',
+            '//cloudresourcemanager.googleapis.com/{}'.format(org_id),
+            self.engine)
+        for org in resources:
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: do you mean access_policy instead of org?

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py [hunk 5]
```diff
@@ -1148,6 +1148,110 @@ def iter_crm_projects(self, parent_type, parent_id):
...
+
+        Yields:
+            dict: Generator of access levels in an organization.
+        """
+        resources = self.dao.iter_cai_assets(
+            ContentTypes.access_level,
+            'cloudresourcemanager.googleapis.com/Organization',
+            access_policy_id,
+            self.engine)
+        for org in resources:
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: access_level instead of org

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py [hunk 6]
```diff
@@ -1148,6 +1148,110 @@ def iter_crm_projects(self, parent_type, parent_id):
...
+
+        Yields:
+            dict: Generator of organization policies for an organization.
+        """
+        resources = self.dao.iter_cai_assets(
+            ContentTypes.org_policy,
+            'cloudresourcemanager.googleapis.com/Organization',
+            '//cloudresourcemanager.googleapis.com/{}'.format(org_id),
+            self.engine)
+        for org in resources:
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: org_policy instead of org

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py [hunk 7]
```diff
@@ -1148,6 +1148,110 @@ def iter_crm_projects(self, parent_type, parent_id):
...
+        Yields:
+            dict: Generator of organization policies for a project.
+        """
+        resources = self.dao.iter_cai_assets(
+            ContentTypes.org_policy,
+            'cloudresourcemanager.googleapis.com/Project',
+            '//cloudresourcemanager.googleapis.com/projects/{}'.format(
+                project_number),
+            self.engine)
+        for org in resources:
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: same as above

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py [hunk 8]
```diff
@@ -1148,6 +1148,110 @@ def iter_crm_projects(self, parent_type, parent_id):
...
+
+        Yields:
+            dict: Generator of organization policies for a folder.
+        """
+        resources = self.dao.iter_cai_assets(
+            ContentTypes.org_policy,
+            'cloudresourcemanager.googleapis.com/Folder',
+            '//cloudresourcemanager.googleapis.com/{}'.format(folder_id),
+            self.engine)
+        for folder in resources:
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: same as above

File: google/cloud/forseti/services/inventory/base/gcp.py [hunk 9]
```diff
@@ -1517,6 +1541,42 @@ def iter_compute_address(self, project_number):
...
+    def iter_crm_organization_access_levels(self, access_policy_id):
+        """Iterate Access Policies from GCP API.
+
+        Args:
+            access_policy_id (str): id of the policy.
+
+        Raises:
+            ResourceNotSupported: Raised for all calls using this class.
+        """
+        raise ResourceNotSupported('Compute AccessLevels are not supported '
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/gcp.py]: access levels is not part of compute.

File: google/cloud/forseti/services/inventory/base/gcp.py [hunk 10]
```diff
@@ -1517,6 +1541,42 @@ def iter_compute_address(self, project_number):
...
+    def fetch_crm_organization_service_perimeter(self, access_policy_id):
+        """Iterate Access Policies from GCP API.
+
+        Args:
+            access_policy_id (str): id of the policy.
+
+        Raises:
+            ResourceNotSupported: Raised for all calls using this class.
+        """
+        raise ResourceNotSupported('Compute ServicePerimeters are not '
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/gcp.py]: same as above

File: google/cloud/forseti/services/inventory/base/gcp.py [hunk 11]
```diff
@@ -1517,6 +1541,42 @@ def iter_compute_address(self, project_number):
...
+    def iter_crm_organization_access_policies(self, org_id):
+        """Iterate Access Policies from GCP API.
+
+        Args:
+            org_id (str): id of the organization.
+
+        Raises:
+            ResourceNotSupported: Raised for all calls using this class.
+        """
+        raise ResourceNotSupported('Compute AccessPolicies are not supported '
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/gcp.py]: same as above

File: google/cloud/forseti/services/inventory/base/resources.py [hunk 12]
```diff
@@ -702,6 +722,29 @@ def key(self):
         return self['name'].split('/', 1)[-1]
 
 
+class ResourceManagerAccessPolicy(resource_class_factory('crm_access_policy',
+                                                         None)):
+    """The Resource implementation for Resource Manager Access Policy."""
+
+    def key(self):
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/resources.py]: Any reason why you cant use the following?

```
class ResourceManagerAccessPolicy(resource_class_factory('crm_access_policy',
                                                         'name')):
```

File: google/cloud/forseti/services/inventory/base/resources.py [hunk 13]
```diff
@@ -711,9 +754,14 @@ def key(self):
         Returns:
             str: key of this resource
         """
-        unique_key = '/'.join([self.parent().type(),
-                               self.parent().key(),
-                               self['constraint']])
+        if 'constraint' not in self._data:
+            unique_key = '/'.join([self.parent().type(),
+                                   self.parent().key(),
+                                   self[0]['constraint']])
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/resources.py]: IIRC CAI provides all constraints per resource in a list format, I don't think adding the first constraint `self[0]['constraint']` to the id is a good idea. I would recommend trying to creating multiple constraints on a resource (e.g. project) and inspect the data generated by CAI. if it's only one row, then parent type and parent key is enough to identify the resource.

File: google/cloud/forseti/services/inventory/base/resources.py [hunk 14]
```diff
@@ -1053,6 +1101,42 @@ def get_iam_policy(self, client=None):
             self.add_warning(err_msg)
             return None
 
+    @cached('org_policy')
+    def get_org_policy(self, client=None):
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/resources.py]: This method doesn't belong to the BQ class

`class BigqueryDataSet:...`

File: google/cloud/forseti/services/inventory/base/resources.py [hunk 15]
```diff
@@ -1053,6 +1101,42 @@ def get_iam_policy(self, client=None):
...
+        try:
+            data, _ = client.iter_crm_organization_org_policies(self['name'])
+            return data
+        except (api_errors.ApiExecutionError, ResourceNotSupported) as e:
+            LOGGER.warning('Could not get Org policy: %s', e)
+            self.add_warning(e)
+            return None
+
+    @cached('access_policy')
+    def get_access_policy(self, client=None):
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/resources.py]: same as above

File: google/cloud/forseti/services/inventory/base/resources.py [hunk 16]
```diff
@@ -1821,12 +1905,29 @@ def iter(self):
                 except ResourceNotSupported as e:
                     # API client doesn't support this resource, ignore.
                     LOGGER.debug(e)
+                except Exception as e:  # pylint: disable=broad-except
```
Comment (joecheuk) [inline google/cloud/forseti/services/inventory/base/resources.py]: why is this needed?

File: google/cloud/forseti/services/model/importer/importer.py [hunk 17]
```diff
@@ -716,15 +723,53 @@ def _convert_resource(self, resource, cached=False, display_key='name',
             type_name=type_name,
             name=resource.get_resource_id(),
             type=resource.get_resource_type(),
-            display_name=data.get(display_key, ''),
-            email=data.get(email_key, ''),
+            display_name=display_name or data.get(display_key, ''),
+            email=data.get(email_key, '') if isinstance(data, dict) else '',
```
Comment (joecheuk) [inline google/cloud/forseti/services/model/importer/importer.py]: please put comment on why these changes are needed

File: tests/services/inventory/test_data/additional_cai_access_policies.dump [hunk 18]
```diff
@@ -0,0 +1,7 @@
+{"name":"//cloudresourcemanager.googleapis.com/organizations/111222333","asset_type":"cloudresourcemanager.googleapis.com/Organization","access_policy":{"name":"accessPolicies/111222333444","parent":"organizations/abc","title":"default policy"},"ancestors":["organizations/111222333"]}
```
Comment (joecheuk) [inline tests/services/inventory/test_data/additional_cai_access_policies.dump]: I dont think this file is being used.

File: tests/services/inventory/test_data/additional_cai_org_policies.dump [hunk 19]
```diff
@@ -0,0 +1,3 @@
+{"name":"//cloudresourcemanager.googleapis.com/organizations/111222333","asset_type":"cloudresourcemanager.googleapis.com/Organization","org_policy":[{"constraint":"constraints/compute.disableSerialPortAccess","update_time":{"seconds":1539891911,"nanos":215000000},"boolean_policy":{"enforced":true}}],"ancestors":["organizations/111222333"]}
```
Comment (joecheuk) [inline tests/services/inventory/test_data/additional_cai_org_policies.dump]: Same here, is this file being used?

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 3 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Done

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 4 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Yeah updated

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 5 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Updated

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 6 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Updated

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 7 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Updated

File: google/cloud/forseti/services/inventory/base/cai_gcp_client.py (hunk 8 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/cai_gcp_client.py]: Updated

File: google/cloud/forseti/services/inventory/base/gcp.py (hunk 9 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/gcp.py]: Yeah updated

File: google/cloud/forseti/services/inventory/base/gcp.py (hunk 10 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/gcp.py]: Updated

File: google/cloud/forseti/services/inventory/base/gcp.py (hunk 11 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/gcp.py]: Updated

File: tests/services/inventory/test_data/additional_cai_org_policies.dump (hunk 19 above)
Comment (red2k18) [reply tests/services/inventory/test_data/additional_cai_org_policies.dump]: Removed file. Mock files are being used.

File: tests/services/inventory/test_data/additional_cai_access_policies.dump (hunk 18 above)
Comment (red2k18) [reply tests/services/inventory/test_data/additional_cai_access_policies.dump]: Removed file. Mock files are being used.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 16 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: Broad except is needed to catch all exceptions other than ResourceNotSupported while iterating through the org policies.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 12 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: `name` for access_policy is not unique. For access_policy it is `//cloudresourcemanager.googleapis.com/organizations/111222333`. Hence `name` cannot be used. Just like `ResourceManagerOrgPolicy()`, for `ResourceManagerAccessPolicy()` we need to use an empty string or `None`.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 13 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: Do you mean creating multiple org policies?

File: google/cloud/forseti/services/model/importer/importer.py (hunk 17 above)
Comment (red2k18) [reply google/cloud/forseti/services/model/importer/importer.py]: Done

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 15 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: Moved it under the resource implementation of Resource Manager Access Policy.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 14 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: Moved it under the resource implementation of Resource Manager Organization Policy.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 13 above)
Comment (joecheuk) [reply google/cloud/forseti/services/inventory/base/resources.py]: org policy constraints.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 14 above)
Comment (joecheuk) [reply google/cloud/forseti/services/inventory/base/resources.py]: You can't get an org policy from an org policy, this method if needed should fall under, organization, folder and project.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 15 above)
Comment (joecheuk) [reply google/cloud/forseti/services/inventory/base/resources.py]: Same here, this should fall under organization

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 16 above)
Comment (joecheuk) [reply google/cloud/forseti/services/inventory/base/resources.py]: This was not here before, we do have other exception handling on a higher level.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 12 above)
Comment (joecheuk) [reply google/cloud/forseti/services/inventory/base/resources.py]: The key method you implemented, is to return the name field. I don't see how that is different than passing in 'name' as the 2nd parameter.

Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: Updated

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 14 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: Done

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 15 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: Done

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 16 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: I later realized this and had uncommented the broad exception. Removed now.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 13 above)
Comment (red2k18) [reply google/cloud/forseti/services/inventory/base/resources.py]: Just verified. Each constraint is retrieved in a row. Added comment in the code.

File: tests/services/inventory/test_data/mock_cai_access_policies.dump [hunk 20]
```diff
@@ -0,0 +1,7 @@
+{"name":"//cloudresourcemanager.googleapis.com/organizations/111222333","asset_type":"cloudresourcemanager.googleapis.com/Organization","access_policy":{"name":"accessPolicies/111222333444","parent":"organizations/abc","title":"default policy"},"ancestors":["organizations/111222333"]}
+{"name":"//cloudresourcemanager.googleapis.com/organizations/111222333","asset_type":"cloudresourcemanager.googleapis.com/Organization","access_policy":{"name":"accessPolicies/111222","parent":"organizations/abc","title":"default policy"},"ancestors":["organizations/111222333"]}
```
Comment (joecheuk) [inline tests/services/inventory/test_data/mock_cai_access_policies.dump]: you can only have one access policy per org

Comment (red2k18) [reply tests/services/inventory/test_data/mock_cai_access_policies.dump]: Yea it doesn't allow creation of second access policy in the same org. Removed second access policy and updated tests accordingly.
